File: src/commonMain/kotlin/com/dropbear/components/Camera.kt
package com.dropbear.components

import com.dropbear.EntityId
import com.dropbear.ecs.Component
import com.dropbear.ecs.ComponentType
import com.dropbear.math.Vector3d

/**
 * Describes a 3D camera, as defined in `dropbear_engine::camera::Camera`
 */
class Camera(
    internal val entity: EntityId,
): Component(entity, "Camera3D") {
    /**
     * The eye/position of the camera.
     */
    var eye: Vector3d
        get() = getCameraEye(entity)
        set(value) = setCameraEye(entity, value)

    /**
     * The target of the entity/the direction it is looking at.
     */
    var target: Vector3d
        get() = getCameraTarget(entity)
        set(value) = setCameraTarget(entity, value)

    /**
     * The up value.
     *
     * Default: [Vector3d.up]
     */
    var up: Vector3d
        get() = getCameraUp(entity)
        set(value) = setCameraUp(entity, value)

    /**
     * The aspect ratio of the camera.
     *
     * Often this is set from the window size. It cannot be overridden.
     */
    val aspect: Double
        get() = getCameraAspect(entity)

    /**
     * The horizontal FOV value.
     */
    var fovY: Double
        get() = getCameraFovY(entity)
        set(value) = setCameraFovY(entity, value)

    /**
     * The nearest distance the camera is able to see.
     *
     * Default: `0.1`
     */
    var znear: Double
        get() = getCameraZNear(entity)
        set(value) = setCameraZNear(entity, value)

    /**
     * The nearest distance the camera is able to see.
     *
     * Default: `100.0`
     */
    var zfar: Double
        get() = getCameraZFar(entity)
        set(value) = setCameraZFar(entity, value)

    /**
     * The horizontal rotational angle.
     *
     * # Calculation
     * ```rust
     * let dir = (builder.target - builder.eye).normalize();
     * let yaw = dir.z.atan2(dir.x);
     * ```
     */
    var yaw: Double
        get() = getCameraYaw(entity)
        set(value) = setCameraYaw(entity, value)

    /**
     * The vertical rotational angle.
     *
     * # Calculation
     * ```rust
     * let dir = (builder.target - builder.eye).normalize();
     * let pitch = dir.y.clamp(-1.0, 1.0).asin();
     * ```
     */
    var pitch: Double
        get() = getCameraPitch(entity)
        set(value) = setCameraPitch(entity, value)

    /**
     * The movement speed of the camera.
     *
     * Default: `1.0`
     */
    var speed: Double
        get() = getCameraSpeed(entity)
        set(value) = setCameraSpeed(entity, value)

    /**
     * The sensitivity of the mouse.
     *
     * Default: `0.002`
     */
    var sensitivity: Double
        get() = getCameraSensitivity(entity)
        set(value) = setCameraSensitivity(entity, value)

    override fun toString(): String {
        return "Camera component of entity $entity \n" +
                "eye: $eye\n" +
                "target: $target\n" +
                "up: $up\n " +
                "aspect: $aspect \n" +
                "fov_y: $fovY \n" +
                "znear: $znear \n" +
                "zfar: $zfar \n" +
                "yaw: $yaw \n" +
                "pitch: $pitch" +
                "speed: $speed" +
                "sensitivity: $sensitivity"
    }

    companion object : ComponentType<Camera> {
        override fun get(entityId: EntityId): Camera? {
            return if (cameraExistsForEntity(entityId)) Camera(entityId) else null
        }
    }
}

expect fun Camera.getCameraEye(entity: EntityId): Vector3d
expect fun Camera.setCameraEye(entity: EntityId, value: Vector3d)
expect fun Camera.getCameraTarget(entity: EntityId): Vector3d
expect fun Camera.setCameraTarget(entity: EntityId, value: Vector3d)
expect fun Camera.getCameraUp(entity: EntityId): Vector3d
expect fun Camera.setCameraUp(entity: EntityId, value: Vector3d)
expect fun Camera.getCameraAspect(entity: EntityId): Double
expect fun Camera.getCameraFovY(entity: EntityId): Double
expect fun Camera.setCameraFovY(entity: EntityId, value: Double)
expect fun Camera.getCameraZNear(entity: EntityId): Double
expect fun Camera.setCameraZNear(entity: EntityId, value: Double)
expect fun Camera.getCameraZFar(entity: EntityId): Double
expect fun Camera.setCameraZFar(entity: EntityId, value: Double)
expect fun Camera.getCameraYaw(entity: EntityId): Double
expect fun Camera.setCameraYaw(entity: EntityId, value: Double)
expect fun Camera.getCameraPitch(entity: EntityId): Double
expect fun Camera.setCameraPitch(entity: EntityId, value: Double)
expect fun Camera.getCameraSpeed(entity: EntityId): Double
expect fun Camera.setCameraSpeed(entity: EntityId, value: Double)
expect fun Camera.getCameraSensitivity(entity: EntityId): Double
expect fun Camera.setCameraSensitivity(entity: EntityId, value: Double)

expect fun cameraExistsForEntity(entity: EntityId): Boolean

File: crates/eucalyptus-exports/src/camera.rs
use eucalyptus_core::ptr::WorldPtr;
use eucalyptus_core::scripting::result::DropbearNativeResult;
use eucalyptus_core::types::NVector3;
use dropbear_engine::camera::Camera;

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "cameraExistsForEntity"),
    c
)]
fn exists_for_entity(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<bool> {
    Ok(world.get::<&Camera>(entity).is_ok())
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraEye"),
    c
)]
fn get_eye(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<NVector3> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.eye.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraEye"),
    c
)]
fn set_eye(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    eye: &NVector3,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.eye = (*eye).into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraTarget"),
    c
)]
fn get_target(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<NVector3> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.target.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraTarget"),
    c
)]
fn set_target(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    target: &NVector3,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.target = target.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraUp"),
    c
)]
fn get_up(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<NVector3> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.up.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraUp"),
    c
)]
fn set_up(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    up: &NVector3,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.up = up.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraAspect"),
    c
)]
fn get_aspect(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.aspect.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraFovY"),
    c
)]
fn get_fovy(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.settings.fov_y.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraFovY"),
    c
)]
fn set_fovy(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    fovy: f64,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.settings.fov_y = fovy.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraZNear"),
    c
)]
fn get_znear(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.znear.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraZNear"),
    c
)]
fn set_znear(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    znear: f64,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.znear = znear.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraZFar"),
    c
)]
fn get_zfar(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.zfar.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraZFar"),
    c
)]
fn set_zfar(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    zfar: f64,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.zfar = zfar.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraYaw"),
    c
)]
fn get_yaw(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.yaw.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraYaw"),
    c
)]
fn set_yaw(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    yaw: f64,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.yaw = yaw.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraPitch"),
    c
)]
fn get_pitch(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.pitch.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraPitch"),
    c
)]
fn set_pitch(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    pitch: f64,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.pitch = pitch.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraSpeed"),
    c
)]
fn get_speed(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.settings.speed.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraSpeed"),
    c
)]
fn set_speed(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    speed: f64,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.settings.speed = speed.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "getCameraSensitivity"),
    c
)]
fn get_sensitivity(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
) -> DropbearNativeResult<f64> {
    match world.get::<&Camera>(entity) {
        Ok(camera) => Ok(camera.settings.sensitivity.into()),
        Err(e) => Err(e.into()),
    }
}

#[dropbear_macro::export(
    kotlin(class = "com.dropbear.components.CameraNative", func = "setCameraSensitivity"),
    c
)]
fn set_sensitivity(
    #[dropbear_macro::define(WorldPtr)] world: &hecs::World,
    #[dropbear_macro::entity] entity: hecs::Entity,
    sensitivity: f64,
) -> DropbearNativeResult<()> {
    match world.get::<&mut Camera>(entity) {
        Ok(mut camera) => {
            camera.settings.sensitivity = sensitivity.into();
            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}
